Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs

```diff
@@ -0,0 +1,252 @@
+                
+                default:
+                    throw new InvalidOperationException(String.Format("Error creating credential from JSON. Unrecognized credential type {0}.", credentialParameters.Type));
+            }
+        }
+
+        internal static UserCredential CreateUserCredentialFromJson(JsonCredentialParameters credentialParameters)
+        {
+            if (credentialParameters.Type != JsonCredentialParameters.AuthorizedUserCredentialType ||
+                string.IsNullOrEmpty(credentialParameters.ClientId) ||
+                string.IsNullOrEmpty(credentialParameters.ClientSecret))
+            {
+                throw new ArgumentException("JSON data do not represent a valid user credential.", "credentialParameters");
+            }
+
+            var initializer = new GoogleAuthorizationCodeFlow.Initializer
+            {
+                ClientSecrets = new ClientSecrets
+                {
+                    ClientId = credentialParameters.ClientId,
+                    ClientSecret = credentialParameters.ClientSecret
+                }
+            };
+
+            var token = new TokenResponse()
+            {
+                RefreshToken = credentialParameters.RefreshToken
+            };
+
+            var flow = new GoogleAuthorizationCodeFlow(initializer);
+            return new UserCredential(flow, "ApplicationDefaultCredentials", token);
+        }
+
+        internal static ServiceAccountCredential CreateServiceAccountCredentialFromJson(JsonCredentialParameters credentialParameters)
+        {
+            if (credentialParameters.Type != JsonCredentialParameters.ServiceAccountCredentialType ||
+                string.IsNullOrEmpty(credentialParameters.ClientEmail) ||
+                string.IsNullOrEmpty(credentialParameters.PrivateKey))
+            {
+                throw new ArgumentException("JSON data do not represent a valid service account credential.", "credentialParameters");
+            }
+            var initializer = new ServiceAccountCredential.Initializer(credentialParameters.ClientEmail);
+            return new ServiceAccountCredential(initializer.FromPrivateKey(credentialParameters.PrivateKey));
+        }
+
+        /// <summary> 
+        /// Helper method to return well known credential file path. This file is created by 
+        /// <a href="https://cloud.google.com/sdk/gcloud/reference/auth/login">gcloud auth login</a>
+        /// </summary>
+        internal string GetWellKnownCredentialFilePath()
+        {
+            return Path.Combine(GetEnvironmentVariable(AppdataEnvironmentVariable) ?? "",
+                CloudSDKConfigDirectory,
+                WellKnownCredentialsFile);
+        }
+
+        /// <summary>Helper method to get the environment variable. Can be overriden for testing.</summary>
+        protected virtual string GetEnvironmentVariable(string variableName)
```

Comment:
I would rephrase it. something like:
Gets the environment variable. This method is protected so it could be overriden for testing purposes only.

How does that sound? Same change to line 246?